### PR TITLE
Global Min IV/Level Filter and Image Based Pokemon Filters

### PR DESCRIPTION
**Pull request by hammydown4325**
**Global Min IV**
The basic structure i went with was adding two fields. The first field was for minIV and the second field was exclude from minIV. I found people wanted to be able to see all tyrans, and drags and such even when miniv was on. 

I have tested this with both Rocketmaps and Monkey Monocle and both work as intended. When the minIV is triggered all mons below the reported IV will not come into new raw_data request. I tried to reload the mons off the IV on the Javascript side but kept crashing tabs in chrome because its a lot more processing than just excluding a mon.

The min IV field is a number field with a minimum number of 0 and a maximum of 100. If the min IV is set to 0 or "" all mons on the map will be displayed regardless of whether they have an IV.

I also found the lots of my members wanted to see rare mons still even if the IV was lower than the global filter. So for this I added the exclude from min iv field. Any mon in this field will display on the map regardless of IV.

My users have enjoyed this feature for a couple weeks now while I was working the kinks out so please give it a try and if someone can test on Base Monocle that would be wonderful.

**Image Based Pokemon Filtering**
Select2 is not what it used to be in the internet world and does not make it easy to quickly filter various pokemon. So I created a simple element that encapsulates all pokemon images into a scrollable easy to deselect filter. This is implemented for exclude_min_iv, hide_pokemon and notify_pokemon.

In the config file I added two variables for this one. The first is $noImageSelect. This allows you to turn on and off the image based filtering and fallback to select 2 if you want. The second is $pathToImages which allows you to define a custom path to where the pokemon images are located. Default path is to the icons-safe directory.

This has been tested on Chrome, Firefox, Safari, and Edge.

### CODE REVIEW

**Comment by hammydown4325** [comment]
Still needs a little work to load in local storage. Should have finished up soon.

File: config/default.php
Comment (123FLO321) [inline config/default.php]: `minIV` not `MinIV`

File: config/example.config.php
Comment (123FLO321) [inline config/example.config.php]: same here

**Comment by hammydown4325** [comment]
I'll change that when I get home. Thanks for or pointing that out.

**Comment by 123FLO321** [comment]
A few thoughts:

- Maybe a red transparent background in addition to the red border would be nice
- Maybe change the border color based on what it does (e.g. exclude red, notify blue, include green)
- Maybe add the same selector used for `min iv` to `notify iv` and `notify level`

And a minor bug:

- In Safari it still opens the picker view

Otherwise really awsome work.

**Comment by hammydown4325** [comment]
Good ideas on the transparency and changing of the colors. I'll today that up tonight thank!!

**Comment by hammydown4325** [comment]
Ok i need to redo some data structuring and stuff to fix the safari bug as its not an easy one to fix. I have the other changes done and will keep you updated of my progress. I hope to have a solution early today.

**Comment by hammydown4325** [comment]
In order to fix this dilema we need to convert the select lists on those three options into input[type="text"] fields. This will stop the select bar from popping up. My question is I originally left falllbacks in place in the code to allow people to continue using select 2 if they didn't want to use Image select.

Because of how input fields in select2 work  each time the data is access and re put in we have to use .join and .split in order to make it compatible. We would need to add a conditional in each of the places where this exists to check if select2 or the other select is used. Do we want to do this or just convert all of PMSF into the new image based select menus? I'll have the polished PR in the morning but wanted to pry some minds overnight. Thanks!

**Comment by Falaris01** [comment]
The change in how the pokemon are selected for notification / excluding is HUGE. 

It has to be the number one complain by users.

I'm all for getting rid of the old way it's done, it was only functional with FireFox anyway.

**Comment by hammydown4325** [comment]
I have made all requested updates to this point.

**Comment by 123FLO321** [comment]
There is a new bug:
Every time a pokemon gets added to the map it plays a sound (it sounds are enabled)

**Comment by hammydown4325** [comment]
Let me take a look at that.

**Comment by hammydown4325** [comment]
@123FLO321 I am not getting this bug when mons are added to the map? Can you export your settings file your using and upload it here so i can have the exact settings. thanks! 

UPDATE: I tried with notify off on all mons and sounds on and nothing played. Tried adding in weedle and each time a weedle spawned it made the weedle noise. Plenty of other mons were spawning that weren't making noises.

**Comment by 123FLO321** [comment]
It sends notifications again and again for a pokemon in the notify list.

**Comment by hammydown4325** [comment]
@123FLO321 Provided me his setting file and found the bug. This is now fixed with this latest commit.

**Comment by peterchabs** [comment]
Nice work d

**Comment by thermoscookies** [comment]
This sounds like an amazing addition! Can't wait to try it out. One question I have, and sorry if this isn't the right place to put it, I am working on a PR to make iv 0's bounce because some of my users like hunting 0 iv's. But of course that would require them to not have the mon excluded. Would it be possible to include this somehow in your global iv filter for =0  and minumum iv rather than having to set minimum >0? This would allow people to show all 0's, 100's and whatever they have on the exclude from min iv list. 

Great work and sorry again if my request is out of line from what this PR is about.

**Comment by hammydown4325** [comment]
Yes it is let me look at your PR and see what I can do

**Comment by mfaalk** [comment]
When Min IV is set to 100 no pokémons shows.
If Min IV is set to 99, then 100 does show.

When adding pokémons to "Exclude from Min IV" I have to move away and then back in order for the pokémons to show up.

Over all, great PR that my users will LOVE!

**Comment by hammydown4325** [comment]
@mfaalk Please test it again and let me know. Thanks for reporting those bugs.

I also converted images to use a sprite sheet which saves 3.2MB on load and should limit the issues Glennmen had with images not fully loading.

**Comment by Acadia1337** [comment]
Anyone have a screenshot of this? I'm curious to see what it looks like.

**Comment by hammydown4325** [comment]
2018-01-08 19_43_31-pmsf glennmen

**Comment by hammydown4325** [comment]
Instead of having two separate pokemon selectors for hide/exclude, I am thinking of wrapping it in some sort of data model toggle that allows you to toggle between two on one but that might be a stretch for this pr at this point as we have changed a lot.

**Comment by hammydown4325** [comment]
FYI for those who dont use the safe icons the sprite sheet should be 1350 x 868. DM me on discord if you need a copy of a sprite sheet.

**Comment by Acadia1337** [comment]
I have something similar set up but I use a button in the top bar "filters" and a modal with the icons.

**Comment by hammydown4325** [comment]
That will most likely come with the complete frontend rewrite that will be happening. I thought about it but at that point we might as well just rewrite the whole frontend to make it more efficient and look less like rocketmaps

**Comment by hammydown4325** [comment]
New Bookmarklet:

`javascript:Store.set('spritefile', '[URL]'); Store.set('spritefileNav', '[URL]');Store.set('spritefileLarge', '[URL]');redrawPokemon(mapData.pokemons);
`

I also forgot to add i fixed a bug with the All/None buttons not properly working.

**Comment by peterchabs** [comment]
everything working great here, 10/10 would merge.

**Comment by hammydown4325** [comment]
Removed Nav Spritesheet after discussions with someone about using background-size to handle the sizing of the bigger sprite sheet. Thus 1 less spritesheet to load.

**Comment by hammydown4325** [comment]
Looking into reported issues with Min IV query on PostGre

**Comment by hammydown4325** [comment]
We figured out the issue with Postgre and I will be making a commit today around lunch time.

**Comment by hammydown4325** [comment]
Min Level Filtering has been added.
UI Interface has been changed a little to make it look nicer for pokemon selection. Now uses tabs to switch between Hide Pokemon and Excl Min IV/Lvl. I also increased the size of the pokemon filter box after many requests for it. 

Test it out and give me feedback. Thanks! 
2018-01-10 20_27_36-pmsf glennmen

**Comment by mfaalk** [comment]
Getting these logs now, nothing showing up on map. (I've modified my PMSF)
image

**Comment by hammydown4325** [comment]
@mfaalk pull the latest commit will fix your issue.

Issues i'm working on.
- Issues with Gym/Raid $noGyms and $noRaids not working to hide them properly. Hide Gym/Raid Sidebar if set to true..
- Add pokemon numbers as little icon on pokemon. Make this toggleable.

**Comment by hammydown4325** [comment]
Current Bug with Rocketmaps levels

**Comment by hammydown4325** [comment]
Added Pokemon Numbers and a config to turn them on and off.
Fixed Rocketmaps MinLvl Bug Tested on Stock Rocketmaps
Made it so that when you set $noPokemon/$noGym/$noRaids to true it hides the associated tabs. Both Gym/Raid need to be hidden in order for it to hide.

**Comment by thermoscookies** [comment]
Not sure if this is an issue or the intended behavior. I have 'Hide Pokemon' set to none. Min IV 90 Min Lvl 20. Then under Notification tab I have 'Notification of Perfection' set to 100. So then I have a 100% level 29 bouncing which is correct but then theres a 100% level 18 that isn't showing at all because it is under the Min Lvl 20 set in the Pokémon tab. 

Should the notification tab 'notify of perfection' override the Min Lvl so that I could make sure I see all 100's bouncing no matter the level or is the way its working, where it will hide a notify of perfection if it is below the Min Lvl the right way?

Thanks again for your fantastic work!

**Comment by hammydown4325** [comment]
Great question. The min level and min Iv currnetly work together in the pr meaning u will only see mons with min IV of 90+ AND min lvl of 20+.  The problem we are having is if we do what ur asking then we literally will have to load every mon onto the site. We could add a toggle that says display 100iv mons always and then ensure they are in the query. This is the exact reason I name the PR Global because the filter is done across all mons and other settings. Hopefully that makes sense to you. Always open to more feedback on it.

**Comment by peterchabs** [comment]
You need to exclude those mons from min level/IV, intended behavior.

**Comment by thermoscookies** [comment]
@iroken222 that does make sense, thank you. so that means my 0iv pr wouldn't work with this because it will always be below the set min iv.

Since you have the global min iv, I think that solves the 'notify of perfection' text box input so maybe we can switch over to having a 'Notify on 100 iv' button instead (and maybe also a 'Notify on 0 iv' :) button) that will do what you said. If those are set to on then include the 100 or 0 iv in the query.

**Comment by hammydown4325** [comment]
I agree and disagree with that at the same time. We are getting into quite a complex level of querying and such with this that we don't want to slow down the sites beyond necessary functionality and such. If we do this sort of thing its needs to be in a logical way that makes absolute sense to the user.  Some people could easily get confused if they see a mon under the min iv and don't realize why.

I think we can wait and revisit this as soon as the pr gets merged.

With that at this point i have heard no more feedback from any other users so currently this pr is completed and ready for review. @CalamityJames @Glennmen 

Once merged I will open up a new PR to rework some of the frontend interface rather than leaving this beast grow beyond a PR at this point.

**Comment by hammydown4325** [comment]
At the request of someone we have renamed the scanner to Monocle_Alternate. I asked the individual if they wanted it removed and they said "Yes Please".

Sorry for an extra commit after the final but wanted to work this in here.

**Comment by LucianoNeo** [comment]
Maaann, first of all CONGRATULATIONS for this awesome job, is much much better than the original (thys system of selection and the incredible IV Filter), I loved and I want to use right now, but there's only one thing not working for me, the weather system (icons cloudy and boosted by) Do you have intention to put this into this PR too ?

**Comment by hammydown4325** [comment]
There have been some questions on what to call the new branch name. I settled on Alternate as to not use anyone names directly. Just a personal opinion with everything going on recently we need to protect individuals as best we can. Feel free to leave an opinion. Thanks!

**Comment by Falaris01** [comment]
Would be nice if you set a Pokemon to be excluded from Min IV it would remove it from the hide Pokemon list. Currently you'd have to go through both lists :)

Or maybe even only having one pokemon list for everything? Clicking the icon will cycle through the possibilities, like e.g. Red = Hide, Blue = Exclude Min IV/Lvl, Green = Exclude MinIv/Lvl and notify? That way the user would only have to go through one list.

Also changing it to MinIV or MinLvl would be prefered for our users I guess. So they can set it up to only see 100IV or lvl35 Pokemon

**Comment by hammydown4325** [comment]
We can probably work in the fact that if they check it on the exclude list it removes it from the hide pokemon list.

I think the singular interface would be too complex for the average user to understand in that way. I considered doing that before but i was getting too confused with colors during my trials. The new interface we would look at would have one giant list of mons with toggles to the right of them for iv, lvl, hide, and notify thus it will allow them to quickly toggle things on and off.  This will most likely come in a future PR for the frontend rewrite. 

We could probably add a toggle field that sets AND or OR for the query that runs that. I have users that want to see mons both 100iv and lvl 35. If we switched it to or that wouldnt be able to do that. so probably adding a field underneath the min iv/ min lvl would be best for that.

**Comment by 123FLO321** [comment]
I found a bug:
If Exl. min IV/Lvl is None it doesn't load any Pokemon

**Comment by hammydown4325** [comment]
@123FLO321 Im not able to reproduce that using Monocle Alternate or RM. I removed all exclude mons and pokemon are loading just fine for me? Can you give me some steps to reproduce and provide your settings export for quicker testing.

2018-01-15 19_30_53-pmsf glennmen

2018-01-15 19_32_51-developer tools - https___dtamap com_map_

**Comment by Glennmen** [comment]
Might be just me but I have a question about PR #52 
Is it added into this PR?
If not, does it still work with this PR?

**Comment by hammydown4325** [comment]
Glennmen yes it works in a sense as long as the Pokemon is visible on the map. He was querying me about extending the pr to support his pr even if min Iv is set. My biggest concern with all of this is the more we add to the database query the longer Everytime it will take to load. Basically we would need to add an OR is 0iv to the get function if they have that toggled on and those mons would have to be in the dataset. Do we display those mons on the map? Idk on that one cause min Iv technically is set so it might be completely fusing for people. I'm open to suggestions.

File: .eslintrc.json
Comment (M4d40) [inline .eslintrc.json]: There is a `,` missing at the end of line 104

Comment (hammydown4325) [reply .eslintrc.json]: Correct in the process of merging develop back into my branch and fixing the merge conflicts.

**Comment by hammydown4325** [comment]
Travis being cranky today lol!!!

**Review by Glennmen — CHANGES_REQUESTED** [review]
Please remove the spritesheet

**Review by Glennmen — APPROVED** [review]

File: lib/RocketMap.php
Comment (CalamityJames) [inline lib/RocketMap.php]: As RocketMap_Sloppy has not been updated, this throws the following warning:
```
PHP Warning:  Declaration of Scanner\RocketMap_Sloppy::get_active($eids, $swLat, $swLng, $neLat, $neLng, $tstamp = 0, $oSwLat = 0, $oSwLng = 0, $oNeLat = 0, $oNeLng = 0) should be compatible with Scanner\RocketMap::get_active($eids, $miniv, $minlevel, $exminiv, $swLat, $swLng, $neLat, $neLng, $tstamp = 0, $oSwLat = 0, $oSwLng = 0, $oNeLat = 0, $oNeLng = 0) 
```
I'd also request swapping to camelCase to keep in the styling of the other vars (change `$minlevel` to `$minLevel` for example). I understand they aren't all in camelCase but going forward we should be doing this, it also makes reading the variables a little easier :)

File: raw_data.php
Comment (CalamityJames) [inline raw_data.php]: Same comment about camelCase here :)

**Review by CalamityJames — CHANGES_REQUESTED** [review]
This is a vast PR, and from what I've tested in a real-world environment very well executed.

I just have a few minor niggles (and a PHP warning) that  need changing.

Caveat: I've mainly been looking at the PHP, and I think it definitely needs another set of eyes or two to look over it, but it's been running in production on my server for a few days with no issues.

File: package.json
Comment (Glennmen) [inline package.json]: I haven't done an in depth code review yet, just a quick question.
Why did you move these dependencies to devDependencies?

Will probably be my small npm knowledge that is failing here.

File: pre-index.php
Comment (TyrolMap) [inline pre-index.php]: We should probably remove the tabs completely instead of only removing their content (for `noHidePokemon` and `noExcludeMinIV`

Comment (123FLO321) [inline pre-index.php]: We should probably remove the tabs completely instead of only removing their content (for `noHidePokemon` and `noExcludeMinIV`)

File: package.json
Comment (hammydown4325) [reply package.json]: That was on accident though we shjould definitely look through these. dependencies are required to run the application where as develop dependencies are only needed to develop. Things like phplint and other things that dont need to be in the build process shouldn't be in there. I'll move them back for now and we can discuss at a later time.

File: lib/RocketMap.php
Comment (hammydown4325) [reply lib/RocketMap.php]: I cannot get the rocketmaps sloppy  error to appear. I dont use rocketmaps sloppy nor do i have a database for it. I have display_errors on and and Display all errors and still nothing appear with either Monocle or Rocketmaps. My guess is it's missing the constructor but thats just a guess from the code and no testing. I'll take care of the camel casing.

Comment (CalamityJames) [reply lib/RocketMap.php]: That's strange, because all libraries are loaded by default. The reason it throws the error is because `get_active()` for RocketMap_Sloppy has a different list of arguments when compared to RocketMap (because you've added `$miniv, $minlevel, $exminiv,`).

I'll take another look in the morning and see if this is not just something exclusive to my test build.

Comment (hammydown4325) [reply lib/RocketMap.php]: Thats weird because rocketmap_sloppy is basically an empty object in all sense other than it inherits rocketmaps. I had one other guy on rocketmaps get the same error but other than that other rocketmaps instances were running fine.

**Comment by hammydown4325** [comment]
I have patched all issues mentioned above except the one @CalamityJames is going to look at.  Please double check and make sure I got all the internationalizations as there are a lot of them. I also found some additional strings like Disable and such and converted them too. @CalamityJames @Glennmen @123FLO321

File: lib/RocketMap.php
Comment (CalamityJames) [reply lib/RocketMap.php]: Entirely my fault @iroken222. My sincere apologies, it was an error on my end. I'll approve these changes now!

**Review by CalamityJames — APPROVED** [review]

File: config/default.php
Comment (Glennmen) [inline config/default.php]: Can you add this to `example.config.php` and better alignment of the comments

File: lib/Monocle.php
Comment (Glennmen) [inline lib/Monocle.php]: why not use `$float` here?

File: lib/Monocle_Alternate.php
Comment (Glennmen) [inline lib/Monocle_Alternate.php]: Monocle/Alternate fork now also has `get_active_by_id()`

File: lib/RocketMap.php
Comment (Glennmen) [inline lib/RocketMap.php]: This seems to be in the wrong order, `$ids` should be the first parameter

File: lib/Monocle.php
Comment (Glennmen) [inline lib/Monocle.php]: ZeChrales fork doesn't have `level` in DB

**Review by Glennmen — CHANGES_REQUESTED** [review]
Please look into these issues.
I only did a code review, not tested it.

File: config/default.php
Comment (hammydown4325) [reply config/default.php]: this is removed. now. Must of missed during cleanup

File: lib/RocketMap.php
Comment (hammydown4325) [reply lib/RocketMap.php]: Good catch updated this

File: lib/Monocle_Alternate.php
Comment (hammydown4325) [reply lib/Monocle_Alternate.php]: Duplicated this function into alternate

File: lib/Monocle.php
Comment (hammydown4325) [reply lib/Monocle.php]: Removed the level portion of ZeChrales

File: lib/Monocle.php
Comment (hammydown4325) [reply lib/Monocle.php]: Missed update. Fixed now

File: lib/Monocle_Alternate.php
Comment (Glennmen) [inline lib/Monocle_Alternate.php]: Not sure where you copied this from but this line has `weight` on develop branch

**Review by Glennmen — CHANGES_REQUESTED** [review]

File: lib/Monocle_Alternate.php
Comment (Glennmen) [inline lib/Monocle_Alternate.php:71]: I think something went wrong when solving the merge conflicts, `$miniv, $minlevel, $exminiv` are removed from function params.

**Review by Glennmen — CHANGES_REQUESTED** [review]

**Review by Glennmen — APPROVED** [review]

**Review by CalamityJames — APPROVED** [review]
Ship it. https://media.giphy.com/media/143vPc6b08locw/giphy.gif

**Pull request merged**
